add bias correction using neigboring stations as reference and clean up existing filter function and notebooks

lepetersson opened the pull request:
This pull request aims to:

- [x] Implement bias correction functionality
- [x] clean up the merged_filter notebook and move the caluclation of `nbrs_not_nan` and `reference` variables to source code

Commit:
first commit

Commit:
move init of soflags to src

Commit:
move calc of nbrs_not_nan & ref to src

Commit:
update notebook

Commit:
remove functionality to set -1 up to the first time step

Commit:
split into 3 separate notebooks

Commit:
clean up notebooks

Commit:
rename notebook

Commit:
add bias corr functionality

Commit:
WIP separate bias corr

Commit:
update index.md

Commit:
skip station if no nbrs

Commit:
improve bias corr

Commit:
clean up bias corr notebook

Commit:
minor fix in notebooks

Commit:
bug fix SO filter

Commit:
fix default variables

Commit:
update test array

Commit:
add bias corr test

cchwala reviewed (commented):
Thanks for the updates. I mainly looked at `flagging.py` and have some minor suggestions there. I did not check the "scientific correctness" of the code. I just looked at code structure.

One point regarding the notebooks. Could you add some plots to see the results of the filters, ideally also with a brief explanation of what can be seen.

lepetersson committed:
Merge branch 'OpenSenseAction:main' into bias_corr

Commit:
fix hooks 1

Commit:
merge

Commit:
fix linting 2

Commit:
fix linting 3

Commit:
fix linting 4

cchwala committed:
trying to fix readthedocs error

lepetersson commented:
> Thanks for the updates. I mainly looked at `flagging.py` and have some minor suggestions there. I did not check the "scientific correctness" of the code. I just looked at code structure.
> 
> One point regarding the notebooks. Could you add some plots to see the results of the filters, ideally also with a brief explanation of what can be seen.

Unfortunately I do not have time to add examples to the notebooks right now (as @JochenSeidel has done nicely with the other notebooks) but I might be able to add something when I am working on the OpenMRG2 and WG2 papers.

Commit:
optimize code

Commit:
add function to calc reference and nbrs not nan

Commit:
minor fix bias corr

Commit:
merge

Commit:
fix linting #5#

cchwala commented:
Code looks good now! Thanks for adding the changes based on my suggestions.

If there is no time now, the notebooks can stay as they are. Extending them later should not be a big problem.

I will try to look into the remaining small CI and readthedocs errors.

cchwala committed:
fix liniting errors (mostly via autfix by ruff)

cchwala committed:
...forgot one notebook...

cchwala committed:
fix problem (which was detected by pylint) with returning potentially non-existing variable `ds_bias_corr`

I hope it is meant to be like I now do it, assuming that `_calc_bias_corr_factor` updates the ds_pws it gets as input and then returns this ds_pws with some additional variables.

cchwala committed:
ignore pylint errors for upper case variable names

cchwala committed:
I ran the bias correction notebook. Let's see if this changes something with the readthedocs build problem...

cchwala committed:
ran the other three notebooks, SO filter notebook interupted on last cells because takes too long, but now let"s see if readthedocs likes the notebooks

cchwala commented:
CI linting now fails because if the error output in the notebooks, but the readthedocs build now works :tada:.

Something with the notebooks where all output was cleared seemed to have caused this problem with readthedocs.

cchwala commented:
@lepetersson some clarification regarding my last comment here. To see the failing test in CI you have to check the CI runs from commit [99c91c9](https://github.com/OpenSenseAction/pypwsqc/pull/46/commits/99c91c916f15cbf802f4693a4ddbef1f1dd26ffe) above. 

Here is a direct link to the failing tests: https://github.com/OpenSenseAction/pypwsqc/actions/runs/18049352931/job/51367908508

The problem seems to be that you changed the number of function arguments in the filter functions, hence you get errors like this in the current tests

```python
TypeError: hi_filter() missing 2 required positional arguments: 'distance_matrix' and 'max_distance'
```

cchwala commented:
If I understand the updated function correctly you only need `distance_matrix` and `max_distance` if you do not have `reference` in `ds_pws`. Hence, you could have `distance_matrix=None` and `max_distance=None` as default and only require them if `reference` (and `nbrs_not_nan`?) is not in `ds_pws`. This way your old tests should still run. But then you need to check also for the case where `distance_matrix` and `max_distance` are passed as args.

Commit:
fix tests

Commit:
fix tests 2

Commit:
fix tests 3

Commit:
fix linting

Commit:
clear output of all cells

Commit:
add output to cells

Commit:
slice timeseries in example use

Commit:
update tests

Commit:
fix linting

Commit:
fix linting

Commit:
update tests

Commit:
handle no nbrs

Commit:
fix error

Commit:
commit

Commit:
fix linting

Commit:
fix test
